File: urbanairship-core/src/test/java/com/urbanairship/PendingResultTest.java
/* Copyright Airship and Contributors */

package com.urbanairship;

import android.os.Looper;
import androidx.annotation.Nullable;

import org.junit.Before;
import org.junit.Test;
import org.robolectric.Shadows;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;

public class PendingResultTest extends BaseTestCase {

    private TestResultCallback<Boolean> resultCallback;
    private PendingResult<Boolean> pendingResult;

    private Looper looper;

    @Before
    public void setup() {
        looper = Looper.myLooper();
        resultCallback = new TestResultCallback<>();
        pendingResult = new PendingResult<>();
        pendingResult.addResultCallback(Looper.myLooper(), resultCallback);
    }

    @Test
    public void testGet() throws ExecutionException, InterruptedException {
        setResultAsync(true);

        // Verify we get the result
        assertTrue(pendingResult.get());
        assertTrue(pendingResult.isDone());
        assertFalse(pendingResult.isCancelled());

        // Verify callback is still called
        Shadows.shadowOf(looper).runToEndOfTasks();
        assertEquals(1, resultCallback.results.size());
        assertTrue(resultCallback.results.get(0));

        // Verify we can still get the result
        assertTrue(pendingResult.get());
    }

    @Test
    public void testCancel() throws ExecutionException, InterruptedException {
        pendingResult.cancel();
        assertTrue(pendingResult.isDone());
        assertTrue(pendingResult.isCancelled());

        setResultAsync(true);

        // Verify the result in get is null
        assertNull(pendingResult.get());

        // Callback should not be called
        assertEquals(0, resultCallback.results.size());
    }

    private void setResultAsync(final boolean result) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                pendingResult.setResult(result);
            }
        }).start();
    }

    private class TestResultCallback<T> implements ResultCallback<T> {

        List<T> results = new ArrayList<>();

        @Override
        public void onResult(@Nullable T result) {
            this.results.add(result);
        }

    }

}

File: urbanairship-core/src/test/java/com/urbanairship/PendingResultTest.kt
/* Copyright Airship and Contributors */
package com.urbanairship

import android.os.Looper
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Shadows

@RunWith(AndroidJUnit4::class)
class PendingResultTest {

    private var resultCallback = TestResultCallback<Boolean>()
    private var pendingResult = PendingResult<Boolean>()

    private var looper = Looper.myLooper()

    @Before
    fun setup() {
        pendingResult.addResultCallback(Looper.myLooper(), resultCallback)
    }

    @Test
    fun testGet() {
        setResultAsync(true)

        // Verify we get the result
        Assert.assertTrue(pendingResult.get() == true)
        Assert.assertTrue(pendingResult.isDone)
        Assert.assertFalse(pendingResult.isCancelled)

        // Verify callback is still called
        Shadows.shadowOf(looper).runToEndOfTasks()
        Assert.assertEquals(1, resultCallback.results.size)
        Assert.assertTrue(resultCallback.results[0])

        // Verify we can still get the result
        Assert.assertTrue(pendingResult.get() == true)
    }

    @Test
    fun testCancel() {
        pendingResult.cancel()
        Assert.assertTrue(pendingResult.isDone)
        Assert.assertTrue(pendingResult.isCancelled)

        setResultAsync(true)

        // Verify the result in get is null
        Assert.assertNull(pendingResult.get())

        // Callback should not be called
        Assert.assertEquals(0, resultCallback.results.size)
    }

    private fun setResultAsync(result: Boolean) {
        Thread { pendingResult.setResult(result) }.start()
    }

    private inner class TestResultCallback<T> : ResultCallback<T> {

        var results = mutableListOf<T>()

        override fun onResult(result: T?) {
            result?.let { results.add(it) }
        }
    }
}
